### DISCUSSION TITLE
Is Longhorn really lightweight block storage solution? v1 burning one core per node :/

### DISCUSSION THREAD

**Discussion by dominch** [longhorn/longhorn#8373] [Q&A]
There are several bugs in github about longhorn high cpu usage, but I'm not sure if I want to talk about bug or is it feature of this system. Also I think that this may be better place for discussion that bug reports. 

According to docs: 
> Longhorn is a lightweight, reliable, and powerful distributed block storage system for Kubernetes.

My context is: small home cluster with 5 nodes (1 master), most are power efficient 4 cores nucs, with 8-64G of RAM, all SSD, few nvmes, 1G ethernet, all debian bookworm x86. There is nothing deployed with super high usage, most things is just idle, used to test interesting services with not much workload. Right now there are about 55 volumes. Longhorn 1.6.1, k3s 1.28.6
Recently I noticed that most of CPU usage comes from longhorn. 
Nothing bad in logs, except 1 or two nodes with errors about wrong iscsi targets (this is already reported, happens all the time on some nodes). 

![image](https://github.com/longhorn/longhorn/assets/10815109/68dd3ed6-d253-4087-b6bd-2e9ca7a30cb9)

I was sure that most of work on cluster is due to prometheus stack, for sure there is plenty of work to done there, it also uses biggest volume 120G.,  It's still way less than longhorn takes. I can see same cpu usage in htop on nodes (about 30% of each). 
This leads to question - is mostly idle cluster requires so much CPU just for longhorn? I'm sure that there is nothing in background tasks (like backup, snapshots or recovery). I already reduced replicas for prometheus (to 2) and maybe I'll setup this bit differently (there is no real data there, I can always prune there and collect new). 

For now I'm looking for debug - why it's taking so much cpu, how to reduce that and how to choose in future wisely. 
As I said - cluster is mostly idle, not that much IO except prometheus.

**Comment by dominch** [comment]
Yet another graph to show the problem: 
![image](https://github.com/longhorn/longhorn/assets/10815109/74ed3c59-a81d-48b3-a6b8-d6bab02aa343)

For now most of computing power is used just by longhorn, 
nothing rebuilds, not much network traffic, no backup in progress or anything big, 

I think this has changed about version 1.4, I see lot of bug reports about high cpu usage, some by specific operations or bugs, some already resolved, I also saw on documentation that longhorn uses quite much resources for querying for some data, with v2 volumes it also uses at least one core for instance manager. 

If there is no way to reduce this then it's clearly not lightweight anymore :/

**Comment by derekbit** [comment]
What are the workloads of the volumes? Are they IO intensive?

**Reply by dominch** [reply]
Nothing is IO intensive, almost everything should be idle for most of time. Please take a look at this graph: 

![image](https://github.com/longhorn/longhorn/assets/10815109/91d58918-f553-4f08-ad22-804a0b821cd7)

Few services are running there smoothly (like smtp/imap, few static https) but nothing should write that much. Most of volumes are empty, the only one that has some data (stored mostly incrementaly) is prometheus, for now it's configured to keep 3 months and it collects about 110G of data at that period. There is also docker image proxy, but this one should not change that much, it keeps maybe 20G of data.
For now there are 53 volumes, without docker proxy, mail and prom stack I think it's less than 20GB of data. I could not track anything that logs too much or doing something in background.

**Comment by dominch** [comment]
With 1.6 and v2 engine there is information about high CPU usage: 
![image](https://github.com/longhorn/longhorn/assets/10815109/828d53a3-e8e6-4a91-a97c-95e297234b3f)

As far as I see I don't have v2 volumes yet, 
Will that consume yet another extra 1 core on each node?

**Comment by dominch** [comment]
There is some hope for v2 engines: https://github.com/longhorn/longhorn/issues/7064  :) :) :) 
@derekbit do You think this usage is some kind of bug or is it normal for v1 and curent version? 
I upgraded to latest version v1.6.2 hoping for any change. Still longhorn burns about one core on each node, RAM usage is more stable thanks to memory leak bug fix, cpu usage is still quite high :/

**Comment by dominch** [comment]
Is this the same issue as described here? : 
https://github.com/longhorn/longhorn/issues/8496

If so, this should be critical error to fix, hopefully this month release will be available soon (with this particular fix)

**Reply by derekbit** [reply]
It is a regression during v1.7.0 development and doesn't happen in v1.6 or before.

**Reply by dominch** [reply]
Ok, 
so still no hope for any change regarding my high cpu usage, 
I already removed 1/3 of loghorn volumes and CPU idel usage is even higher :/
